File: problems/栈与队列/84.柱状图中最大的矩形.py
#!/usr/bin/env python3
"""
84. 柱状图中最大的矩形
算法: 单调递增栈
时间复杂度: O(n)
空间复杂度: O(n)
"""
def largest_rectangle_area(heights: list[int]) -> int:
    stack = []
    max_area = 0
    for i, h in enumerate(heights + [0]):
        while stack and heights[stack[-1]] > h:
            height = heights[stack.pop()]
            width = i if not stack else i - stack[-1] - 1
            max_area = max(max_area, height * width)
        stack.append(i)
    return max_area


if __name__ == "__main__":
    print(largest_rectangle_area([2, 1, 5, 6, 2, 3]))  # 预期: 10
    print(largest_rectangle_area([2, 4]))               # 预期: 4
    print(largest_rectangle_area([1]))                  # 预期: 1

File: problems/栈与队列/84.柱状图中最大的矩形.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>
using namespace std;

int largestRectangleArea(vector<int>& heights) {
    stack<int> st;
    int max_area = 0;
    vector<int> h = heights;
    h.push_back(0);
    for (int i = 0; i < h.size(); i++) {
        while (!st.empty() && h[st.top()] > h[i]) {
            int height = h[st.top()]; st.pop();
            int width = st.empty() ? i : i - st.top() - 1;
            max_area = max(max_area, height * width);
        }
        st.push(i);
    }
    return max_area;
}

int main() {
    // 测试用例
    vector<int> h1 = {2, 1, 5, 6, 2, 3};
    cout << largestRectangleArea(h1) << endl;  // 预期: 10

    vector<int> h2 = {2, 4};
    cout << largestRectangleArea(h2) << endl;  // 预期: 4

    vector<int> h3 = {1};
    cout << largestRectangleArea(h3) << endl;  // 预期: 1

    return 0;
}
